### PR TITLE
Api/connection: HCM and Connections support for Codec and MaxHeaders …

### PR DESCRIPTION
**Pull request by nfuden**
# Description

Please include a summary of the changes.

Per #5437 update Connection api to include more Options for HCM as reflected in the original protos.
Note that this does not close the issue as there is 

# Context

Users would like to be able to set the three options noted in the #5437 these include Codec type and max_headers_count. These we could easily pull from dependent libraries without breaking changes. Further info for the the final change requested (max_requests_per_connection) will be bundled in a separate discussion related to go-control plane updates.

# Checklist:

- [ /] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [ /] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [ ] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [ ] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [ /] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ /] I have made corresponding changes to the documentation
- [/ ] I have added tests that prove my fix is effective or that my feature works

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/5437

### CODE REVIEW

File: projects/gloo/api/v1/options/hcm/hcm.proto [hunk 1]
```diff
@@ -115,4 +115,28 @@ message HttpConnectionManagerSettings {
     // The :ref:`http_connection_manager.path_with_escaped_slashes_action_sampling<config_http_conn_man_runtime_path_with_escaped_slashes_action_enabled>` runtime
     // variable can be used to apply the action to a portion of all requests.
     PathWithEscapedSlashesAction path_with_escaped_slashes_action = 26;
+ 
+    // The maximum number of headers. If unconfigured, the default
+    // maximum number of request headers allowed is 100. Requests that exceed this limit will receive
+    // a 431 response for HTTP/1.x and cause a stream reset for HTTP/2.
+    google.protobuf.UInt32Value max_headers_count = 28;
```
Comment (jenshu) [inline projects/gloo/api/v1/options/hcm/hcm.proto:122]: nit: can we move this up to line 78 next to `max_connection_duration` and `max_stream_duration`, as they are all part of the http protocol options? (we can keep the value as `28` but just move this field up)

Comment (nfuden) [reply projects/gloo/api/v1/options/hcm/hcm.proto]: Good call. I also took the opportunity to clean up the existing fields' comments to mirror each other and have the appropriate linkage to envoy docs.

File: projects/gloo/api/v1/options/hcm/hcm.proto [hunk 2]
```diff
@@ -74,7 +74,10 @@ message HttpConnectionManagerSettings {
 
     // For an explanation of these settings see https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/protocol.proto#config-core-v3-httpprotocoloptions
     google.protobuf.Duration max_connection_duration = 23;
+     // For an explanation of these settings see https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/protocol.proto#config-core-v3-httpprotocoloptions
     google.protobuf.Duration max_stream_duration = 24;
+     // For an explanation of these settings see https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/protocol.proto#config-core-v3-httpprotocoloptions
+    google.protobuf.UInt32Value max_headers_count = 28;
```
Comment (jenshu) [inline projects/gloo/api/v1/options/hcm/hcm.proto:80]:
```suggestion
    google.protobuf.UInt32Value max_headers_count = 27;
```

File: projects/gloo/api/v1/options/hcm/hcm.proto [hunk 3]
```diff
@@ -115,4 +118,23 @@ message HttpConnectionManagerSettings {
...
+        // The connection manager will assume that the client is speaking HTTP/1.1.
+        HTTP1 = 1;
+
+        // The connection manager will assume that the client is speaking HTTP/2
+        // (Envoy does not require HTTP/2 to take place over TLS or to use ALPN.
+        // Prior knowledge is allowed).
+        HTTP2 = 2;
+    }
+    // Supplies the type of codec that the connection manager should use.
+    CodecType codec_type = 29;
```
Comment (jenshu) [inline projects/gloo/api/v1/options/hcm/hcm.proto:139]:
```suggestion
    CodecType codec_type = 28; 
```

**Review by jenshu — APPROVED** [review]

**Review by MitchAman — APPROVED** [review]

**Pull request merged**